### PR TITLE
[doc] Update release notes for PMD 7.0.0-rc1

### PR DESCRIPTION
**Pull request by adangel**
Preview: https://github.com/adangel/pmd/blob/pmd7-update-release-notes/docs/pages/release_notes.md
and https://github.com/adangel/pmd/blob/pmd7-update-release-notes/docs/pages/release_notes_pmd7.md

## Describe the PR

The goal is to complete the release notes for PMD 7.0.0-rc1.

- [x] record gif - analyzing openjdk11 - for README.md and release notes
- [x] minimum requirement: java8,
- [x] emoticons for headings, see [kirby release notes](https://github.com/getkirby/kirby/releases) or apache release drafter
  https://github.com/apache/maven-gh-actions-shared/blob/main/.github/release-drafter.yml
- [x] Mention language versions - document, that there might be parse errors. version is to be used for minimum/maximumLanguageVersion in rules
- [x] Review section about language modules / language lifecycle. Remove some details. Link to the howto pages.
  - see https://github.com/pmd/pmd/pull/4060/files#r1096787602
- [x] Mention split of CLI and ANT into own modules
- [x] Section: Notes for integrators
- [x] Incorporate infos from https://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/pmd_next_major_development.html
- :x: more API changes?
  - let's do this with the upcoming RCs to complete the release notes

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No regression tested rules have been changed.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by jsotuyod — COMMENTED** [review]
I think we should also probably explicitly mention that we split away CLI into `pmd-cli` and Ant into `pmd-ant`, which were originally part of `pmd-core`… most integrations are probably relying on either of them (and I'd expect specially on Ant, ie: Gradle)…

Maybe a dedicated section named **Notes for integrators**, reinforcing that, as a major release, this is a major rewrite, and we did these changes, along with removing the old `PMD` class, and explicitly marking which classes are internal and subject to change and which not,  and therefore expect all integrators to require some level of change to adapt, and guide people to look at the deprecations notes for more details.

**Review by oowekyala — COMMENTED** [review]
I just had that pending

File: docs/pages/7_0_0_release_notes.md [hunk 1]
```diff
@@ -84,17 +120,21 @@ To use it, edit your `~/.bashrc` / `~/.zshrc` file and add the following line:
...
 
-Languages backed by an Antlr grammar are now fully supported. This means, it's now possible not only to use Antlr grammars for CPD,
-but we can actually build full-fledged PMD rules for them as well. Both the traditional Java visitor rules, and the simpler
-XPath rules are available to users.
+Languages backed by an [Antlr](https://www.antlr.org/) grammar are now fully supported. This means, it's now
+possible not only to use Antlr grammars for CPD, but we can actually build full-fledged PMD rules for them as well.
+Both the traditional Java visitor rules, and the simpler XPath rules are available to users. This allows
+to leverage existing grammars.
 
 We expect this to enable both our dev team and external contributors to largely extend PMD usage for more languages.
```
Comment (jsotuyod) [inline docs/pages/7_0_0_release_notes.md:133]:
```suggestion

Contributors: [Lucas Soncini](https://github.com/lsoncini) (@lsoncini),
  [Matías Fraga](https://github.com/matifraga) (@matifraga),
  [Tomás De Lucca](https://github.com/tomidelucca) (@tomidelucca)
```

File: docs/pages/7_0_0_release_notes.md [hunk 2]
```diff
@@ -143,30 +190,51 @@ node.descendants(ASTMethodCall.class)
...
 
+Contributors: [Clément Fournier](https://github.com/oowekyala) (@oowekyala)
 
 #### JavaScript support
 
 The JS specific parser options have been removed. The parser now always retains comments and uses version ES6.
 The language module registers only one version (as before), now correctly with version "ES6" instead of "3".
-Since there is only one version available for JavaScript there is actually no need to selected a specific version.
+Since there is only one version available for JavaScript there is actually no need to select a specific version.
 The default version is always ES6.
```
Comment (jsotuyod) [inline docs/pages/7_0_0_release_notes.md:209]: most of this is no longer true after #4387, in particular https://github.com/pmd/pmd/pull/4387/files#diff-6a73814684c68604f37a965213cbfa5e3c0bbf7f4d5d075de9f2f987682122d3

File: docs/pages/7_0_0_release_notes.md [hunk 3]
```diff
@@ -123,16 +165,21 @@ We are shipping the following rules:
...
+(and only) supported XPath version is now XPath 3.1. This version of the XPath language is mostly identical to
+XPath 2.0.
 
+Notable changes:
+* The deprecated support for sequence-valued attributes is removed. Sequence-valued properties are still supported.
+* Refer to [the Saxonica documentation](https://www.saxonica.com/html/documentation/expressions/xpath31new.html) for
+  an introduction to new features in XPath 3.1.
 
-#### Node stream API
+#### Node stream API for AST traversal
```
Comment (jsotuyod) [inline docs/pages/7_0_0_release_notes.md:182]: I'd probably move this section a little further down, next to API, moving from the more user-focused changes and towards the more technical ones.

File: docs/pages/7_0_0_release_notes.md (hunk 2 above)
Comment (adangel) [reply docs/pages/7_0_0_release_notes.md:209]: That's already addressed by the very same PR (https://github.com/pmd/pmd/pull/4387/files#diff-2fdc02f732cb8c6cc158d3e9cf5c05566e942377d5bd1761b45f9479bf3f811e) - I'll rebase this PR soon, then this section should read differently.

**Comment by adangel** [comment]
FYI - I've created the gif using the following commands:

```
curl -L -o jdk-master.zip https://github.com/openjdk/jdk/archive/refs/heads/master.zip
unzip jdk-master.zip
export PMD_JAVA_OPTS="-Dorg.slf4j.simpleLogger.log.net.sourceforge.pmd=off"
pmd check -R rulesets/java/quickstart.xml -d jdk-master/src/java.base -f text --cache pmd.cache --report-file jdk-report.txt
```

Record screen cast with https://github.com/EasyScreenCast/EasyScreenCast (a gnome3 extension)
Convert webm to gif: https://engineering.giphy.com/how-to-make-gifs-with-ffmpeg/

    ffmpeg -i pmd7-demo.webm -filter_complex "[0:v] fps=12,scale=960:-1,split [a][b];[a] palettegen [p];[b][p] paletteuse" pmd7-demo.gif

File: docs/pages/7_0_0_release_notes.md [hunk 4]
```diff
@@ -17,7 +17,55 @@ This is a {{ site.pmd.release_type }} release.
...
+
+#### New official logo
+
+Many of you probably have already seen the new logo, but now it's time to actually ship it. The new logo
+was long ago decided (see [#1663](https://github.com/pmd/pmd/issues/1663)).
+We decided it's time to have a modernized logo and get rid of the gun. This allows to include
+the logo anywhere without offense.
+
+
+![New PMD Logo](images/logo/pmd-logo-300px.png)
```
Comment (adangel) [inline docs/pages/7_0_0_release_notes.md:39]: We have a couple of relative links which refer to files/pages inside the repo and the generated page. These won't work on e.g. Github Releases.

We probably should replace them with absolute links (https://pmd.github.io/latest/...) - which then only are valid after the release of pmd-7.0.0-rc1.

File: docs/pages/7_0_0_release_notes.md [hunk 5]
```diff
@@ -17,7 +17,55 @@ This is a {{ site.pmd.release_type }} release.
...
+PMD 7 does not need to load classes, and does not have these problems.
+
+The AST exposes much more semantic information now. For instance, you can jump from a method call to 
+the declaration of the method being called, or from a field access to the field declaration. These
+improvements allow interesting rules to be written that need precise knowledge of the types
+in the program, for instance to detect unnecessary boxing or unnecessary explicit type arguments (TODO link those rules).
+These are just a small preview of the new rules we will be adding in the PMD 7 release cycle.
+
+Some first results of the Java AST changes are for now documented in the Wiki:
+[Java clean changes](https://github.com/pmd/pmd/wiki/Java_clean_changes).
```
Comment (adangel) [inline docs/pages/7_0_0_release_notes.md:61]: I think for rc1, we can keep the link to the wiki page. We have #4294 to address this.

File: docs/pages/7_0_0_release_notes.md [hunk 6]
```diff
@@ -17,7 +17,55 @@ This is a {{ site.pmd.release_type }} release.
...
+
+Overall, the changes to the parser, AST, type resolution and symbol table code has made PMD for Java **significantly faster**. On average, we have seen ~2-3X faster analysis, but as usual, this may change depending on your workload, configuration and ruleset.
+TODO: Take infos from <http://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/pmd_next_major_development.html#java>
```
Comment (adangel) [inline docs/pages/7_0_0_release_notes.md:64]: TODO...

File: docs/pages/7_0_0_release_notes.md [hunk 7]
```diff
@@ -123,50 +180,69 @@ We are shipping the following rules:
...
-use cases.
+The API of PMD has been growing over the years and needed some cleanup. The goal is, to
+have a clear separation between a well-defined API and the implementation, which is internal.
+This should help us in future development.
+
+#### Compatibility and migration notes
+
+PMD 7 requires Java 8 or above to execute.
+
+TODO - what to look out for. Hints and known problems.
```
Comment (adangel) [inline docs/pages/7_0_0_release_notes.md:227]: Maybe we can postpone this section, we have #4294 planned...
Or maybe keep it very short

* CLI changed: Custom scripts need to be updated
* Java module revamped: Custom rules need to be updated
* Deleted rules: Custom rulesets need to be reviewed. See below for a list of new and deleted rules.
* XPath 1.0 support is removed, `violationSuppressXPath` now requires XPath 2.0 or 3.1: Custom rulesets need to be reviewed.
* Custom rules using rulechains: some API changes (around `RuleTargetSelector`, `AbstractRule#buildTargetSelector()`)

File: docs/pages/7_0_0_release_notes.md (hunk 4 above)
Comment (adangel) [reply docs/pages/7_0_0_release_notes.md]: I've introduced a page variable:

```
baseurl: https://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/
```

This variable needs to be adjusted.

The ruletags generate a slightly different url:

```
url_prefix = "https://pmd.github.io/pmd-#{context["site.pmd.version"]}/"
```

Maybe that should be aligned? The ruletags url will only work when the release is done and the doc is uploaded to pmd/pmd.github.io.

**Review by adangel — COMMENTED** [review]
I think, we should create two release notes pages: Our usual "release_notes.md" and additionally a "release_notes_PMD7.md".
The reason is, that it becomes longer and longer.

The usual "release_notes.md", which will also be used for the release notes on GitHub Releases, should IMHO be a bit shorted and mention each topic with one or two sentences and point to a more detailed section in the other release note page. It should be concise and just have a pointer to the details, if someone is interested.

The new "release_notes_PMD7.md" would then be longer and contain all the details. The page title could be "Detailed Release Notes for PMD7". E.g. it also contains the list of deprecated APIs we collected from all the PMD6 releases - these are currently in "next_major_development.md".
The page could also serve as a migration guide (#4294) or at least be the basis for such a guide.

File: docs/pages/7_0_0_release_notes.md (hunk 4 above)
Comment (adangel) [reply docs/pages/7_0_0_release_notes.md]: I implemented now a better solution that makes it work for the release notes rendering as well as the site rendering.

For site-rendering, we use relative urls, for release notes, we use the deployed url - similar what rule tags does.

File: docs/pages/7_0_0_release_notes.md (hunk 5 above)
Comment (adangel) [reply docs/pages/7_0_0_release_notes.md]: I started to move it into the detailed release notes page and added a couple of TODOs.

File: docs/pages/7_0_0_release_notes.md (hunk 6 above)
Comment (adangel) [reply docs/pages/7_0_0_release_notes.md]: I've merged pmd_next_major_development with the new detailed pmd_release_notes_pmd7 page. We can add anything that is missing during the course of the upcoming RCs.

**Comment by adangel** [comment]
IMHO this is now ready.

I've made the main release notes a bit smaller and more concise: https://github.com/adangel/pmd/blob/pmd7-update-release-notes/docs/pages/release_notes.md

There's now a second page, which contains the full details: https://github.com/adangel/pmd/blob/pmd7-update-release-notes/docs/pages/release_notes_pmd7.md

This second page was originally the "next_major_development" page - I've merged the contents. It's not fully fledged out yet for Java, but I think, we can finish this with the next RCs.

**Pull request merged**